### ISSUE TITLE
Unexpected low input bandwidth with Oboe vs OpenSL in some configurations

### ISSUE THREAD

**Issue by chrismanchester** [google/oboe#1673]
Android version(s): 10 and 11, possibly others
Android device(s):  Oppo and Moto devices, possibly others
Oboe version: 1.6.1
App name used for testing: Sing!

 
**Short description**
This may be more of a question about how best to configure our stream than a proper bug report, but we're receiving unexpected results with Oboe/AAudio versus our prior OpenSL implementation.

We've received several reports of "muffled voice" recordings from our Oboe implementation. Inspecting the frequency content of the uploads it appears these vocal files had a capture sample rate of about 16k. Users perceive this as a downgrade from our Open SL implementation, where we can see with the same input device the rate was apparently more like 32k.

For Open SL we usually set sample rate according to `AudioManager.PROPERTY_OUTPUT_SAMPLE_RATE` to have the best chance of getting the low latency path.

For Oboe we pass a sample rate (again from `AudioManager.PROPERTY_OUTPUT_SAMPLE_RATE`) and request conversion so the underlying stream sees `oboe::Unspecified`. I recently noticed this comment in `QuirksManager.cpp` about this behavior:

```
    // If a SAMPLE RATE is specified for low latency, let the native code choose an optimal rate.
    // This isn't really a workaround. It is an Oboe feature that is convenient to place here.
    // TODO There may be a problem if the devices supports low latency
    //      at a higher rate than the default.
```

This reminded me of these user reports and I wonder if it's pointing in the general direction of the problem. I have also realized our requirements may be a little different from what we're requesting, and there may be a point where we would favor higher input sample rate over a low latency stream if we can't have both.

Questions:

* Will AAudio sometimes select a lower sample rate than necessary for input? Is there a way to ensure this doesn't occur?
* Is there a way to ensure we get the highest sample rate for a given configuration with the current set of APIs, perhaps by determining the set of input sample rates that would result in a low latency stream so we can select the highest?

**Steps to reproduce**
We have been unable to reproduce this in-house. We see this primarily for wired headset mics.

**Expected behavior**
AAudio provides the highest possible input sample rate for low latency input streams, or a way to specify this behavior.

**Actual behavior**
We sometimes unexpectedly see a low capture sample rate.

**Device**

moto g(8) power lite, CPH1931 (Oppo  A5), others

**Any additional context**

We can compile a more complete list of devices and OS versions if that would be helpful.

**Comment by philburk** [comment]
What is the value of AudioManager.PROPERTY_OUTPUT_SAMPLE_RATE?

I do not recommend using that for Oboe. AAudio can get better information from the HAL.  That was just a hack for OpenSL ES.

> We see this primarily for wired headset mics.

Are those USB or 3.5mm? It would help to know the exact headset model. 
Some USB headsets only record at 44100. Some record up to 192 KHz, which can also be a problem.

According to https://www.gsmarena.com/motorola_moto_g8_power_lite-10154.php
The Moto device uses a Mediatek MT6765. It has a 3.5mm jack.

The Oppo A5 uses a Qualcomm SDM665.

The phones are quite different. So probably not a HAL issue.

>  Inspecting the frequency content of the uploads it appears these vocal files had a capture sample rate of about 16k.

I assume that means the spectrum dropped off at 8K. There may be an issue with an EQ that is getting enabled. Or if the sample rate conversion is active in Oboe then the anti-aliasing filter may be too aggressive.

> We have been unable to reproduce this in-house. 

Is that because you do not have those devices? Or you have the device but they do not sound muffled?

> We can compile a more complete list of devices and OS versions if that would be helpful.

Yes, I would love to test one of those devices. Otherwise we have no idea what is really happening. The more devices you can identify the more likely I can find one in our library.

> Will AAudio sometimes select a lower sample rate than necessary for input?

We do not ask for anything lower than 48000. But we may get something lower depending on the device.

> Is there a way to ensure we get the highest sample rate for a given configuration with the current set of APIs, 
> perhaps by determining the set of input sample rates that would result in a low latency stream so we can
> select the highest?

We are looking at ways to optimize the USB configuration. Currently we choose the highest sample rate. That can be a problem if the headset runs at 192 KHz.

**Comment by chrismanchester** [comment]
Thanks for taking a look! This is certainly not as straightforward as I had hoped. I dug through a lot of user reports today and was finally able to reproduce the issue on a moto test device.

> What is the value of AudioManager.PROPERTY_OUTPUT_SAMPLE_RATE?

This was 48000 in the impacted performances I've seen.

> 
> I do not recommend using that for Oboe. AAudio can get better information from the HAL. That was just a hack for OpenSL ES.
> 
> > We see this primarily for wired headset mics.
> 
> Are those USB or 3.5mm? It would help to know the exact headset model. Some USB headsets only record at 44100. Some record up to 192 KHz, which can also be a problem.

For the Oppo examples these are reported as something like "CPH1911: wired headset" or "h2w: wired headset". For the moto example I'm seeing this with the device microphone.

> 
> According to https://www.gsmarena.com/motorola_moto_g8_power_lite-10154.php The Moto device uses a Mediatek MT6765. It has a 3.5mm jack.
> 
> The Oppo A5 uses a Qualcomm SDM665.
> 
> The phones are quite different. So probably not a HAL issue.
> 
> > Inspecting the frequency content of the uploads it appears these vocal files had a capture sample rate of about 16k.
> 
> I assume that means the spectrum dropped off at 8K. There may be an issue with an EQ that is getting enabled. Or if the sample rate conversion is active in Oboe then the anti-aliasing filter may be too aggressive.

That is the symptom. After looking through more examples I see the cutoff is closer to 6k most of the time.

> 
> > We have been unable to reproduce this in-house.
> 
> Is that because you do not have those devices? Or you have the device but they do not sound muffled?

I have a moto g(20) that reproduced this today, at least in part. The built in mic on this device only appears to support up to 16k. If I use "Record and Play" in OboeTester I get the "full" 8k, but if I make a recording with our duplex stream I have nothing above about 5k. If I turn off sample rate conversion in for our streams and give the output sr to the input stream I once again get content up to 8k.

It seems like we should turn off conversion to avoid this loss, but input and output need to agree since we're using `FullDuplexStream`. Would you recommend passing the output stream sample rate to the input stream, or perhaps vice-versa? Will this be rejected in some circumstances?

> 
> > We can compile a more complete list of devices and OS versions if that would be helpful.
> 
> Yes, I would love to test one of those devices. Otherwise we have no idea what is really happening. The more devices you can identify the more likely I can find one in our library.

Aside from my moto g(20), the CPH1931, CPH1911, and CPH1941 are the only devices I've found so far where I can confirm the pattern of low bandwidth on Oboe/AAudio fixed by moving a user back to OpenSL. "muffled voice" can be a rather generic complaint, but I'll keep sifting through reports. Unfortunately I'm also seeing instances where only certain Oboe performances from the user have the issue so it may not reproduce reliably.

> 
> > Will AAudio sometimes select a lower sample rate than necessary for input?
> 
> We do not ask for anything lower than 48000. But we may get something lower depending on the device.
> 
> > Is there a way to ensure we get the highest sample rate for a given configuration with the current set of APIs,
> > perhaps by determining the set of input sample rates that would result in a low latency stream so we can
> > select the highest?
> 
> We are looking at ways to optimize the USB configuration. Currently we choose the highest sample rate. That can be a problem if the headset runs at 192 KHz.

**Comment by chrismanchester** [comment]
> > > We have been unable to reproduce this in-house.
> > 
> > 
> > Is that because you do not have those devices? Or you have the device but they do not sound muffled?
> 
> I have a moto g(20) that reproduced this today, at least in part. The built in mic on this device only appears to support up to 16k sample rate. If I use "Record and Play" in OboeTester I get the full 8k bandwidth, but if I make a recording with our duplex stream I have nothing above about 5k. If I turn off sample rate conversion in for our streams and give the output sr to the input stream I once again get content up to 8k.
> 
> It seems like we should turn off conversion to avoid this loss, but input and output need to agree since we're using `FullDuplexStream`. Would you recommend passing the output stream sample rate to the input stream, or perhaps vice-versa? Will this be rejected in some circumstances?

We are particularly interested in how best to achieve this while still having the best opportunity to acquire low latency streams.

> 
> > > We can compile a more complete list of devices and OS versions if that would be helpful.
> > 
> > 
> > Yes, I would love to test one of those devices. Otherwise we have no idea what is really happening. The more devices you can identify the more likely I can find one in our library.
> 
> Aside from my moto g(20), the CPH1931, CPH1911, and CPH1941 are the only devices I've found so far where I can confirm the pattern of low bandwidth on Oboe/AAudio fixed by moving a user back to OpenSL. "muffled voice" can be a rather generic complaint, but I'll keep sifting through reports. Unfortunately I'm also seeing instances where only certain Oboe performances from the user have the issue so it may not reproduce reliably.
> 

I was able to find track down two more models exhibiting the issue: the Oppo CPH1969 and the TECNO LD7.

**Comment by philburk** [comment]
> The built in mic on this device [moto g(20)] only appears to support up to 16k sample rate. 

How did you determine that? Is that from Microphone Info reports?
Or from examining the spectrum?

> If I use "Record and Play" in OboeTester I get the full 8k bandwidth, 

What sample rate do you specify when recording? If you select sampleRate=0 then does it open a stream at 16000 Hz?

In "Record and Play" you can enable the sample rate converter in Oboe by selecting from the "SRC:" menu in the middle of the screen.

If you record at 44100 and the native rate is 48000 then an SRC will run somewhere. If you select SRC quality None then it will resample in AudioFlinger. If you select any other quality then Oboe will open a child stream at the native rate, add a sample rate converter, and then add a parent stream at 44100 that reads from the SRC and presents the audio to your app.

    AAudioStream@48000 => oboeStream@48000 => resampler => oboeStream@44100 => app

You can see the Oboe logs using:
    adb logcat -c && adb logcat *:S OboeAudio:D

Look for something like this:

    12-13 08:28:36.131 29393 29393 I OboeAudio: configure() flowgraph converts channels: 2 to 2, format: 2 to 2, rate: 48000 to 44100, cbsize: 0 to 0, qual = 2

That will tell us if the Oboe resampler is active.

> but if I make a recording with our duplex stream I have nothing above about 5k. 

What settings do you use for the INPUT stream?
Specifically sampleRate and [setSampleRateConversionQuality](https://google.github.io/oboe/classoboe_1_1_audio_stream_builder.html#a0c98d21da654da6d197b004d29d8499c)

Note that if the sample rate you request and the native sample rate are the same then Oboe will not add a resampler.

OboeTester has a "DEVICE REPORT" test that prints speaker and mic info.

**Comment by chrismanchester** [comment]
> > The built in mic on this device [moto g(20)] only appears to support up to 16k sample rate.
> 
> How did you determine that? Is that from Microphone Info reports? Or from examining the spectrum?

I suspect this because if I don't select a sample rate in "Record and Play" the sample rate is reported as 16k, and because I can't find a recording configuration that shows spectral content above 8k. According to the device report the built in microphone supports 8000, 11025, 12000, 16000, 22050, 24000, 32000, 44100, and 48000 as sample rates.

> 
> > If I use "Record and Play" in OboeTester I get the full 8k bandwidth,
> 
> What sample rate do you specify when recording? If you select sampleRate=0 then does it open a stream at 16000 Hz?

Yes, exactly.

> 
> In "Record and Play" you can enable the sample rate converter in Oboe by selecting from the "SRC:" menu in the middle of the screen.
> 
> If you record at 44100 and the native rate is 48000 then an SRC will run somewhere. If you select SRC quality None then it will resample in AudioFlinger. If you select any other quality then Oboe will open a child stream at the native rate, add a sample rate converter, and then add a parent stream at 44100 that reads from the SRC and presents the audio to your app.
> 
> ```
> AAudioStream@48000 => oboeStream@48000 => resampler => oboeStream@44100 => app
> ```
> 
> You can see the Oboe logs using: adb logcat -c && adb logcat *:S OboeAudio:D
> 
> Look for something like this:
> 
> ```
> 12-13 08:28:36.131 29393 29393 I OboeAudio: configure() flowgraph converts channels: 2 to 2, format: 2 to 2, rate: 48000 to 44100, cbsize: 0 to 0, qual = 2
> ```
> 
> That will tell us if the Oboe resampler is active.
> 
> > but if I make a recording with our duplex stream I have nothing above about 5k.
> 
> What settings do you use for the INPUT stream? Specifically sampleRate and [setSampleRateConversionQuality](https://google.github.io/oboe/classoboe_1_1_audio_stream_builder.html#a0c98d21da654da6d197b004d29d8499c)
> 
> Note that if the sample rate you request and the native sample rate are the same then Oboe will not add a resampler.

We pass the same sample rate to the input and output streams and use

`builder.setSampleRateConversionQuality(oboe::SampleRateConversionQuality::Medium);`

With these settings I get:

```I/OboeAudio: configure() flowgraph converts channels: 2 to 2, format: 1 to 1, rate: 16000 to 48000, cbsize: 0 to 0, qual = 3```

and the resulting recording is band limited to 5 or maybe 6 khz.

If I omit `setSampleRateConversionQuality`, pass `oboe::Unspecified` to the output, and pass the output stream sample rate to the input I am able to see spectral content up to 8k.

> 
> OboeTester has a "DEVICE REPORT" test that prints speaker and mic info.

I wasn't able to export this info, but I attached a data paths test result that seems to have the relevant details.
[OboeTester-DataPaths-motorola-moto-g(20)-20221214-020040.txt](https://github.com/google/oboe/files/10220840/OboeTester-DataPaths-motorola-moto-g.20.-20221214-020040.txt)

**Comment by philburk** [comment]
Wow! It is all very clear now. Thanks for the additional info. I don't normally see devices that default to 16000 for recording. 

Now it is clear that the "muffling" is due to the Oboe resampler. The sinc based resampling algorithm inherently includes a low pass filter. It appears that needs to be tuned better for upsampling. I will run some frequency response tests. I think we can fix this.

We have another bug that may be related. #1625 

I really appreciate the time you spent to report this in such detail. It is very helpful. Sorry for the trouble.

**Comment by chrismanchester** [comment]
Thanks very much! We will happily pull in a fix whenever available

**Comment by philburk** [comment]
Here is a simple way to demonstrate the problem.

1. Launch OboeTester
2. Tap TEST OUTPUT
3. Select SRate: 12000
4. Select Signal: "White Noise"
5. Select SRC: quality "None" to use AudioFlinger SRC
6. Tap OPEN START
7. Hear bright White Noise
8. Tap CLOSE
9. Select SRC: quality "Best" to use AudioFlinger SRC
10. Tap OPEN START

EXPECT to hear same bright noise

ACTUAL hear muffled noise

**Comment by philburk** [comment]
> We will happily pull in a fix whenever available

Small fix in #1813 being reviewed. Sound fine for white noise and sine waves.
How does it sound for music and voice?

**Comment by chrismanchester** [comment]
> > We will happily pull in a fix whenever available
> 
> Small fix in #1813 being reviewed. Sounds fine for white noise and sine waves. How does it sound for music and voice?

Works great. Thanks!
